editoast/src/modelsv2/projects.rs

```diff
@@ -0,0 +1,265 @@
+        self.update_last_modified_conn(&mut conn).await
+    }
+
+    pub async fn update_last_modified_conn(
+        mut self,
+        conn: &mut PgConnection,
+    ) -> Result<Option<Project>> {
+        self.last_modification = Utc::now().naive_utc();
+        self.save(conn).await?;
+        Ok(Some(self))
+    }
+
+    pub async fn with_studies(self, db_pool: Data<DbPool>) -> Result<ProjectWithStudies> {
+        use crate::tables::study::dsl as study_dsl;
+        let conn = &mut db_pool.get().await?;
+        let studies_count = study_dsl::study
+            .filter(study_dsl::project_id.eq(self.id))
+            .count()
+            .get_result(conn)
+            .await?;
+        Ok(ProjectWithStudies::new_from_project(self, studies_count))
+    }
+
+    pub async fn delete_and_prune_document(
+        conn: &mut PgConnection,
+        project_id: i64,
+    ) -> Result<bool> {
+        let project_obj =
+            Project::retrieve_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
+                .await?;
+        let _ = Project::delete_static(conn, project_id).await?;
+
+        if let Some(image) = project_obj.image {
+            // We don't check the result. We don't want to throw an error if the image is used in another project.
+            let _ = Document::delete_static(conn, image).await;
+        }
+
+        Ok(true)
+    }
+}
+
+pub trait UpdateAndPruneDocument {
+    async fn update_and_prune_document(
+        self,
+        conn: &mut PgConnection,
+        project_id: i64,
+    ) -> Result<Project>;
+}
+
+impl UpdateAndPruneDocument for Changeset<Project> {
+    async fn update_and_prune_document(
+        self,
+        conn: &mut PgConnection,
+        project_id: i64,
+    ) -> Result<Project> {
+        let old_project =
+            Project::retrieve_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
+                .await?;
+        let image_to_delete = if Some(old_project.image) != self.image {
+            old_project.image
+        } else {
+            None
+        };
+        let project = self
+            .update_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
+            .await?;
+        if let Some(image) = image_to_delete {
+            // We don't check the result. We don't want to throw an error if the image is used in another project.
+            let _ = Document::delete_static(conn, image).await;
+        }
+        Ok(project)
+    }
+}
+
+#[async_trait]
+impl List<Ordering> for Project {
+    async fn list_conn(
+        conn: &mut PgConnection,
+        page: i64,
+        page_size: i64,
+        ordering: Ordering,
+    ) -> Result<PaginatedResponse<Self>> {
+        let ordering = ordering.to_sql();
+        let project_rows = sql_query(format!("SELECT t.* FROM project as t ORDER BY {ordering}"))
+            .paginate(page, page_size)
+            .load_and_count::<Row<Project>>(conn)
```

Comment:
```suggestion
            .load_stream(conn)
            .await?
            .map(Self::from_row)
```

Using a stream you'll avoid a useless allocation and benefit from type inference.

Reply:
I got those errors:
```
error[E0609]: no field `results` on type `futures::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:178:14
    |
178 |             .results
    |              ^^^^^^^ unknown field

error[E0609]: no field `count` on type `futures::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:184:33
    |
184 |             count: project_rows.count,
    |                                 ^^^^^ unknown field

error[E0609]: no field `previous` on type `futures::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:185:36
    |
185 |             previous: project_rows.previous,
    |                                    ^^^^^^^^ unknown field

error[E0609]: no field `next` on type `futures::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:186:32
    |
186 |             next: project_rows.next,
    |                                ^^^^ unknown field

For more information about this error, try `rustc --explain E0609`.
warning: `editoast` (bin "editoast" test) generated 1 warning (1 duplicate)
error: could not compile `editoast` (bin "editoast" test) due to 4 previous errors; 1 warning emitted
warning: build failed, waiting for other jobs to finish...
error[E0609]: no field `results` on type `futures_util::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures_util::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:178:14
    |
178 |             .results
    |              ^^^^^^^ unknown field

error[E0609]: no field `count` on type `futures_util::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures_util::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:184:33
    |
184 |             count: project_rows.count,
    |                                 ^^^^^ unknown field

error[E0609]: no field `previous` on type `futures_util::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures_util::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:185:36
    |
185 |             previous: project_rows.previous,
    |                                    ^^^^^^^^ unknown field

error[E0609]: no field `next` on type `futures_util::stream::Map<std::pin::Pin<std::boxed::Box<dyn futures_util::Stream<Item = std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>> + std::marker::Send>>, fn(std::result::Result<diesel_async::pg::row::PgRow, diesel::result::Error>) -> std::result::Result<_, diesel::result::Error>>`
   --> src/modelsv2/projects.rs:186:32
    |
186 |             next: project_rows.next,

```
I'm not sure that it inferes the type

Reply:
It's because `load_stream` will load a `ProjectRow` and not a PaginatedResponse, which you'll have to instantiate manually.

But if it's too annoying to impl it this way, you can leave it as it is, it's fine as well.